GPU implementation (complete)

kartikmandar opened the pull request:
It has some CUDA conflicts right now that I have to sort out. While testing cupy and cupy-cuda11x and using CUDA directly, I have messed up my env that I have to sort it out before testing. I will probably try this on collab/runpod if this problem persists. 

I will add request for review when I feel its error free and tutorial is working fine.

kartikmandar committed:
add gpu tutorial

kartikmandar committed:
Implement GPU backend using cupy and cufinufft

kartikmandar committed:
add setup call in gpu simulate

kartikmandar committed:
some cleaning of code

kartikmandar committed:
correct gpu_nufft_transforms and compare with cpu_nufft_transforms

kartikmandar committed:
Add some tests to compare cpu_beams and gpu_beams

kartikmandar committed:
Add GPU acceleration for beam evaluation

- Implement direct GPU evaluation for AiryBeam and UVBeam interpolation
- Achieve 40-2000x speedup while maintaining numerical accuracy

kartikmandar committed:
replace print by logging.warning

kartikmandar committed:
Add comprehensive four-way visibility simulation tests

- Test fftvis (CPU/GPU) and matvis (CPU/GPU) produce identical results
- Fix GPU implementation slice handling and loop structure
- All implementations match within numerical precision

kartikmandar committed:
Some performance tests for gpu vs cpu

kartikmandar committed:
Add GPU tutorial updates and performance comparison notebook

TODO: Fix GPU-CPU accuracy differences (max diff ~56.3)
TODO: Fix tutorial typos and remove debugging cells

kartikmandar committed:
Fix failing tests for GPU implementation

- Update test_use_gpu_function to handle cached GPU availability correctly
- Fix CPU simulation tests to match current fftvis-matvis API compatibility
- Update GPU tests to use current CuPy API (OutOfMemoryError, random.random)
- Increase Ray object store memory to meet minimum requirements

kartikmandar committed:
Add comprehensive performance benchmark notebook

Compare fftvis (CPU/GPU) vs matvis performance across varying sources, times, frequencies, and baselines

kartikmandar committed:
Merge upstream/main: Update module names and resolve conflicts

- Renamed modules: cpu_beams→beams, cpu_nufft→nufft, etc (keeping cpu_simulate/gpu_simulate)
- Fixed all imports to use new module names
- Resolved conflicts in 4 test files

kartikmandar committed:
Fix API compatibility issues between fftvis and matvis

- Fixed CoordinateRotation.select_chunk() call to use correct number of arguments (2 instead of 3)
- Fixed backend parameter compatibility by separating CPU-specific parameters in wrapper
- Fixed test to use correct beam function names for CPU vs GPU comparison

kartikmandar committed:
Revert select_chunk fix - method requires both chunk and time args

kartikmandar committed:
Fix select_chunk calls in GPU implementation and syntax error

kartikmandar committed:
resolved copilot review suggestions

steven-murray commented on src/fftvis/cli.py:
```diff
@@ -105,8 +104,7 @@ def run_profile(
 
-    # Add profiling to the simulate_vis function and the CPU implementation's _evaluate_vis_chunk
-    profiler.add_function(simulate_vis)
+        profiler.add_function(simulate_vis)
```
```suggestion
    profiler.add_function(simulate_vis)
```

steven-murray commented on src/fftvis/core/antenna_gridding.py:
```diff
@@ -8,21 +8,19 @@ def find_integer_multiplier(
-    """
-    Return the smallest positive integer f such that f * arr[i] is integral
-    for all entries (up to rational approximation). Zeros in arr are ignored.
+    """Find smallest integer to make all array values integral.
```
```suggestion
    """
    Find the smallest integer product that makes all array values integral.
    
    This function ignores any zero-values in `arr`.
```

steven-murray commented on src/fftvis/core/antenna_gridding.py:
```diff
@@ -41,27 +39,23 @@ def can_scale_to_int(
+        Array to check.
     tol : float
-        Tolerance for checking if values are close to integers in units of meters.
+        Tolerance in meters.
```
I prefer the old description for this one.

steven-murray commented on src/fftvis/core/antenna_gridding.py:
```diff
@@ -75,63 +69,45 @@ def find_lattice_basis(
-    differences between the antenna positions and selects two non-collinear
-    vectors to form the basis. If all vectors are collinear, it returns a default
-    basis.
+    """Find lattice basis for antenna positions.
```
Why is the full description deleted?

steven-murray commented on src/fftvis/core/antenna_gridding.py:
```diff
@@ -142,62 +118,43 @@ def check_antpos_griddability(
-    if the positions can be scaled to integers. If they can, it returns the
-    modified antenna positions and the transformation matrix. If not, it returns
-    the original antenna positions and an identity transformation matrix.
+    """Check if antenna positions can be gridded.
```
Again, let's not remove the entire description. Have a short summary sentence, then move the body of hte description to a new paragraph.

steven-murray commented on src/fftvis/core/antenna_gridding.py:
```diff
@@ -142,62 +118,43 @@ def check_antpos_griddability(
-        The transformation matrix used to scale the antenna positions.
-        This is a 3x3 matrix that transforms the original coordinates
-        to the grid coordinates.
+    tuple
+        (is_griddable, modified_antpos, transform_matrix)
```
This is not the correct format for documenting return values... the original is correct.

steven-murray reviewed (dismissed):
Hi @kartikmandar I've reviewed some of the files but I've realized that most of hte changes are to documentation, which I didn't expect from this PR. Some of hte changes are reasonable but others are deleting much more inf than necessary. Let's keep docs changes to a separate PR so we can focus on them better.

kartikmandar commented:
Sure I cherry-picked the newest commit after going to the two before it, I can work on the documentation on a different PR separately.

kartikmandar committed:
Fix GPU beam interpolation issues

- Fix type checking to handle BeamInterface-wrapped beams
- Add polarized flag to cache key to prevent data collision
- Fix CuPy compatibility with wrap mode for azimuth coordinates
- GPU and CPU now match within numerical precision (3e-9)

kartikmandar committed:
Enhance benchmark notebook with extended parameter ranges and results saving

- Add 4-5 more tiers for each benchmark parameter (sources, times, frequencies, antennas)
- Add GPU vs CPU speedup to all speedup comparison plots
- Add system information collection (CPU, RAM, GPU specs)
- Add comprehensive results saving with timestamped directory
- Include visibility consistency checks between all backends
- Fix baseline calculation to match matvis (all N×N baselines)
- Generate summary reports with key findings and speedup statistics

kartikmandar committed:
Remove Type 1+2 fallback from GPU NUFFT, add beta version guidance

Simplified GPU NUFFT fallback hierarchy by removing the incorrect Type 1+2
decomposition. Now throws informative error with installation instructions
for cufinufft==2.4.0b1 and finufft==2.4.0rc1 when Plan-based Type 3 fails.

kartikmandar committed:
Edit some failing tests for now

kartikmandar committed:
Fix CI test failures by adding GPU dependency checks

- Add conftest.py with centralized GPU availability detection
- Skip GPU tests when cupy is not available
- Add fallbacks for optional dependencies (tabulate, matplotlib)
- Fix conditional imports in GPU test modules

kartikmandar commented:
I am focusing on increasing test coverage and writing better tests now.

kartikmandar committed:
Fix GPU performance with frequency batching

GPU was 2.7x slower than CPU for high frequency counts due to sequential
processing. Now batch multiple frequencies per NUFFT call.

- Add batch NUFFT functions for 2D and 3D transforms
- Process frequencies in batches (up to 128) based on GPU memory
- Reduce kernel launches from O(n_freq) to O(n_freq/batch_size)

Results: 240 frequencies now 5.78x faster (79.58s → 13.77s), GPU now
2.5x faster than CPU.

kartikmandar committed:
Improve GPU memory management and benchmark robustness

Added memory_manager.py to centralize GPU memory handling with monitoring and cleanup functions. Enhanced benchmark notebook with memory-aware testing that gracefully handles GPU memory limits. Updated GPU simulation for better memory efficiency and improved frequency batching performance.

kartikmandar committed:
Merge branch 'main' into gpu-implementation

kartikmandar committed:
Merge branch 'gpu-implementation' of https://github.com/kartikmandar/fftvis into gpu-implementation

kartikmandar committed:
Add gpu tests and cufinufft tests

steven-murray commented on .gitignore:
```diff
@@ -169,4 +169,12 @@ memray-*.html
+# Benchmark output directories
+docs/tutorials/A100_highRam_benchmark/
+docs/tutorials/QuadroP600_highRam_benchmark/
+docs/tutorials/T4_highRam_benchmark/
```
Maybe make these like `docs/tutorials/*_benchmark/`

kartikmandar replied:
sure.

steven-murray commented on .gitignore:
```diff
@@ -169,4 +169,12 @@ memray-*.html
+docs/tutorials/T4_highRam_benchmark/
+
+# Local testing against matvis
+matvis/
```
This seems strange to have. If you're doing ad-hoc testing against matvis, perhaps do the testing outside the repo.

kartikmandar replied:
Shifted to outside repo.

steven-murray commented on pyproject.toml:
```diff
@@ -9,10 +9,10 @@ name = "fftvis"
-    {name="Tyler Cox", email='tyler.a.cox@berkeley.edu'},
-    {name="Steven Murray", email="steven.murray@sns.it"}
+    { name = "Tyler Cox", email = 'tyler.a.cox@berkeley.edu' },
+    { name = "Steven Murray", email = "steven.murray@sns.it" },
```
Add yourself!

kartikmandar replied:
Done.

steven-murray commented on src/fftvis/__init__.py:
```diff
@@ -13,6 +13,16 @@
+    _gpu_available = True
+except ImportError:
+    _gpu_available = False
+    GPUBeamEvaluator = None
```
I wonder if a better implementation here would be to define each of the `GPUBeamEvaluator` and `GPUSimulationEngine` as dummy functions, like:

```python
def GPUBeamEvaluator(*args, **kwargs):
    raise ImportError("GPU dependencies not installed. Install fftvis with the [gpu] extra")

```

kartikmandar replied:
Added a warning for the user.

steven-murray commented on src/fftvis/utils.py:
```diff
@@ -14,18 +16,26 @@
 )
 
+# Cache the GPU availability to avoid repeated checks
+_cached_use_gpu = None
```
Instead of this rather complicated structure where you have a cached variable and a function, you could just have the function code at module-level. Also, I think "use gpu" is a little confusing, because you're not deciding here whether to *use* GPU, just whether it is available to use. So, e.g.:

```python
try:
    import cupy as cp

    # Check if a CUDA device is actually available
    GPU_AVAIL = cp.cuda.is_available()
    if not GPU_AVAIL:
        logging.warning("CuPy installed but no CUDA device found. Using CPU backend.")
       
except ImportError:
    GPU_AVAIL = False
```

Since module imports are cached, this will only run once.

kartikmandar replied:
I simplified this check directly into __init__.py and GPU_available variable is True only when all three are available: CuPy installed, CUDA hardware available, cufinufft installed (with apt warning if false)

steven-murray commented on src/fftvis/gpu/beams.py:
```diff
@@ -2,87 +2,449 @@
         beam: BeamInterface,
-        az: np.ndarray,
-        za: np.ndarray,
+        az: cp.ndarray,
```
I'm a little puzzled as to why we didn't more closely use the beam API from matvis. The implementation here is quite convoluted. Can it be broken up, or made more similar to the matvis implementation, given that matvis already has GPU beams?

steven-murray commented on src/fftvis/gpu/gpu.py:
```diff
@@ -2,23 +2,87 @@
-Currently it contains a stub implementation.
 """
 
+import ray
```
In matvis, the main simulation module is just called "gpu.py". Can we do that here too?

steven-murray commented on src/fftvis/gpu/gpu_simulate.py:
```diff
@@ -51,41 +115,590 @@ def simulate(
+            logger.warning(
+                "GPU backend only supports 'az_za_map_coordinates' for beam interpolation. Ignoring provided interpolation_function."
+            )
+            interpolation_function = "az_za_map_coordinates"
```
Shouldn't this validation be done on the GPUBeamInterpolator object itself?

kartikmandar replied:
removed it.

steven-murray commented on src/fftvis/gpu/gpu_simulate.py:
```diff
@@ -51,41 +115,590 @@ def simulate(
+            if hasattr(beam, "Nfreqs") and beam.Nfreqs > 1:
+                beam = beam.interp(
+                    freq_array=freqs, new_object=True, run_check=False
+                )  # pragma: no cover
```
why the "no cover" here? We should be covering this line

kartikmandar replied:
removed it and a couple of others.

steven-murray commented on src/fftvis/gpu/gpu_simulate.py:
```diff
@@ -51,41 +115,590 @@ def simulate(
+        is_coplanar = np.all(np.less_equal(np.abs(bls_cpu[2]), flat_array_tol))
+
+        # Scale baselines by 1/c (on CPU)
+        bls_cpu /= utils.speed_of_light
```
I'm getting the feeling that there's a LOT of shared code between this and the CPU implementation. Can we re-use code here at all?

steven-murray commented on src/fftvis/gpu/gpu_simulate.py:
```diff
@@ -51,41 +115,590 @@ def simulate(
+                f"Please use CPU backend."
+            )
+            raise ValueError(
+                f"Dataset too large for GPU memory. Use backend='cpu' for {total_sources:,} sources."
```
We should also be able to chunk over the source axis, right? I.e. we should never *have* to use the CPU

steven-murray commented on src/fftvis/gpu/gpu.py:
```diff
@@ -51,41 +115,590 @@ def simulate(
+                )
+        
+        # Optimize source chunk size and frequency batch size together
+        optimal_source_chunk, optimal_freq_batch = mem_manager.optimize_chunking(
+            nsources_total=total_sources,
+            nfreqs_total=nfreqs,
+            ntimes_total=ntimes,
+            nbls=nbls,
+            nfeeds=nfeeds,
+            precision=precision,
+            polarized=polarized,
+            min_chunk_size=1000,
+            max_freq_batch=nfreqs  # Allow processing all frequencies if memory permits
+        )
```
Now I'm confused -- here we're allowing for chunking of sources, but just above we were returning an error when the number of sources was too large for GPU memory. I'm not following the logic here...

steven-murray commented on src/fftvis/gpu/gpu_simulate.py:
```diff
@@ -51,41 +115,590 @@ def simulate(
+        if hasattr(coord_mgr, 'chunk_size'):
+            if coord_mgr.chunk_size is None or coord_mgr.chunk_size > optimal_source_chunk:
+                coord_mgr.chunk_size = optimal_source_chunk
+                logger.info(f"Updated coordinate manager chunk size to {optimal_source_chunk:,}")
```
Why not just do this logic of getting the chunk size before setting up the coord manager? Instead of mutating the object later? It's almost always a bad idea to mutate an object in place.

steven-murray commented on src/fftvis/gpu/gpu_simulate.py:
```diff
@@ -51,41 +115,590 @@ def simulate(
+        rotation_matrix_gpu = cp.asarray(rotation_matrix_cpu)
+        bls_gpu = cp.asarray(bls_cpu)
+        freqs_gpu = cp.asarray(freqs)
+        # Beam data transfer is handled by GPUBeamEvaluator internally
```
This should just be handled in the coord manager, just like in matvis

steven-murray commented on src/fftvis/gpu/gpu_simulate.py:
```diff
@@ -51,41 +115,590 @@ def simulate(
+        # nthreads is ignored for GPU NUFFT, but keep the loop structure
+        nthreads_per_proc = [
+            1
+        ] * nprocesses  # Each GPU process typically uses 1 thread for NUFFT
```
Put the comment on the line above so the formatting looks nicer for this line

kartikmandar replied:
done

steven-murray commented on src/fftvis/gpu/gpu_simulate.py:
```diff
@@ -51,41 +115,590 @@ def simulate(
+                fnc(
+                    time_idx=tc,
+                    freq_idx=fc,
+                    beam=beam_ref,  # Pass Ray object ref or direct object
```
Probably don't need the many repeated comments, just one

kartikmandar replied:
done

steven-murray commented on src/fftvis/gpu/gpu_simulate.py:
```diff
@@ -51,41 +115,590 @@ def simulate(
-        rotation_matrix: np.ndarray,
-        bls: np.ndarray,
-        freqs: np.ndarray,
+        beam,  # UVBeam or BeamInterface (on CPU, needs data transfer)
```
I think it's better to always expect a BeamInterface -- that's the reason they exist, so they can bridge the gap between analytic and UVBeams

steven-murray commented on src/fftvis/gpu/gpu_simulate.py:
```diff
@@ -51,41 +115,590 @@ def simulate(
+        if not isinstance(bls, cp.ndarray):
+            bls = cp.asarray(bls)
+        if not isinstance(freqs, cp.ndarray):
+            freqs = cp.asarray(freqs)
```
To make the code less complex, I would not do this. This function is private, so it should only ever be called by the main program, which should have the correct logic in it to convert the arrays. If you still think it's a good idea, you can anyway make this less complex by simply doing the `.asarray()` for each of the arrays -- it's a no-op when the array is already a cupy array (it returns a view).

kartikmandar replied:
removed the checks and ensured only cupy arrays are passed.

steven-murray commented on src/fftvis/gpu/gpu_simulate.py:
```diff
@@ -51,41 +115,590 @@ def simulate(
+                        if not isinstance(topo, cp.ndarray):
+                            topo = cp.asarray(topo)
+                        if not isinstance(flux_sqrt, cp.ndarray):
+                            flux_sqrt = cp.asarray(flux_sqrt)
```
Instead, just trust that `coord_mgr` is doing the right thing... again, *you* control what's going on in this function. It should only ever be using a GPU-ified coord manager, right?

kartikmandar replied:
Yes, removed the check.

steven-murray commented on src/fftvis/gpu/gpu_simulate.py:
```diff
@@ -51,41 +115,590 @@ def simulate(
+                    if not isinstance(az, cp.ndarray):
+                        az = cp.asarray(az)
+                    if not isinstance(za, cp.ndarray):
+                        za = cp.asarray(za)
```
again, don't have to do this because we control the call above

kartikmandar replied:
corrected.

steven-murray commented on src/fftvis/gpu/gpu_simulate.py:
```diff
@@ -51,41 +115,590 @@ def simulate(
+                            if flux_sqrt.ndim > 1:
+                                flux_slice = flux_sqrt[:nsim_sources, freq_idx]
+                            else:
+                                flux_slice = flux_sqrt[:nsim_sources]
```
This can be pulled out of the if/else because it's the same code in both branches

kartikmandar replied:
moved out.

steven-murray commented on src/fftvis/gpu/gpu_simulate.py:
```diff
@@ -51,41 +115,590 @@ def simulate(
+                            A_s_reshaped = A_s.reshape(nfeeds**2, nsim_sources)
+                        except ValueError:
+                            logger.warning(f"Cannot reshape A_s with shape {A_s.shape} to {(nfeeds**2, nsim_sources)}")
+                            continue
```
Seems like we should just let it error out here?

kartikmandar replied:
Yeah, raising a valueerror would be better than just a warning.

steven-murray commented on src/fftvis/gpu/gpu_simulate.py:
```diff
@@ -51,41 +115,590 @@ def simulate(
+                        
+                        # Ensure correct dtype
+                        if A_s_reshaped.dtype != complex_dtype:
+                            A_s_reshaped = A_s_reshaped.astype(complex_dtype)
```
Why would it ever not be the correct type?

kartikmandar replied:
The GPubeamevaluator doesn't have a precision parameter. So it accepts whatever dtype it gets from pyuvdata. I should rather make evaluate_beam() respect precision and respect the correct dtype. and remove conversion from here.

steven-murray commented on src/fftvis/gpu/gpu_simulate.py:
```diff
@@ -51,41 +115,590 @@ def simulate(
+                        mempool = cp.get_default_memory_pool()
+                        mempool.free_all_blocks()
+                    except:
+                        pass
```
why? Do bad things happen if the memory isn't cleaned up? In what cases does this happen?

steven-murray commented on src/fftvis/gpu/memory_manager.py:
```diff
@@ -0,0 +1,299 @@
+            'total_per_time': rotation_matrix + coord_rotation_overhead,
+        }
+    
+    def optimize_chunking(
```
I feel like I've seen this code (or wrote this code...) before somewhere but I can't find it. If it does exist elsewhere, is it possible to re-use the existing code?

steven-murray commented on src/fftvis/gpu/nufft.py:
```diff
@@ -1,98 +1,533 @@
+        weights_reshaped = weights
+
+    # Use native Type 3 if available, otherwise use Plan-based fallback
+    if HAS_NATIVE_TYPE3_SUPPORT:
```
In what situation is the native Type 3 not available? It would be great to not have to manage all the fallback situations if we don't have to.

tyler-a-cox replied:
I think this might be leftover from when only the Github version of finufft had support for Type 3 on GPU, but not the published version. I agree though, we probably should just default to supporting Type 3 without the fallback

kartikmandar replied:
Yeah, it was not available earlier, and only the beta version of finufft had support for the type 3, so I had to resort to this fallback. It's properly available in a stable version, so we can remove that now.

steven-murray commented on src/fftvis/gpu/utils.py:
```diff
@@ -1,34 +1,144 @@
 
 
-def inplace_rot(rot: np.ndarray, b: np.ndarray):  # pragma: no cover
+def cleanup_gpu_memory():
```
Do we ever actually use this function? Is it important? I don't love the idea of having an arbitrary 0.1 second pause in here... depending on the context of how this is called, that could add up a lot.

steven-murray commented on src/fftvis/gpu/utils.py:
```diff
@@ -1,34 +1,144 @@
+                logger.error(
+                    f"GPU memory exhausted after {max_retries} retries with reduced batch sizes.\n"
+                    f"This dataset is too large for your GPU memory.\n"
+                    f"Please switch to CPU backend by using backend='cpu' in your simulate_vis() call."
```
I don't love all these instructions to use CPU instead... I don't think there should ever be a realistic case in which GPU can't be used -- we can batch over the source axis as much as we want.

steven-murray commented on src/fftvis/gpu/utils.py:
```diff
@@ -1,34 +1,144 @@
-    # result_gpu = cp.matmul(rot_gpu, b_gpu)
-    # cp.copyto(b, cp.asnumpy(result_gpu))
+    # Ensure inputs are cupy arrays
+    if not isinstance(rot, cp.ndarray):
```
I don't think the if statement is necessary as `cp.asarray` should be a no-op if the argument is already a cupy array

steven-murray commented on src/fftvis/wrapper.py:
```diff
@@ -42,7 +42,13 @@ def create_beam_evaluator(
+        evaluator = GPUBeamEvaluator(**kwargs)
+        # Ensure the beam_list is properly initialized since this is required by matvis
+        evaluator.beam_list = []
+        evaluator.beam_idx = None
```
This should be done in the GPUBeamEvaluator?

steven-murray commented on src/fftvis/wrapper.py:
```diff
@@ -230,29 +235,35 @@ def simulate_vis(
+    # Add CPU-specific parameters
+    if backend == 'cpu':
+        common_params['upsample_factor'] = upsample_factor
+        common_params['force_use_type3'] = force_use_type3
```
For clarity, I recommend doing the following instead:

```python
if backend == 'cpu':
    backend_params = {'upsample_factor': upsample_factor, 'force_use_type3': force_use_type3}
elif backend == 'gpu':
    backend_params = {}

return engine.simulate(**(common_params | backend_params))
```

kartikmandar committed:
change gitignore benchmark files to glob pattern

kartikmandar committed:
Move matvis to outside repo for random testing

kartikmandar committed:
added myself(kartik) as maintainer

kartikmandar committed:
add helpful warning for user when calling gpu beam evaluator without installing dependencies

kartikmandar committed:
GPU_available check in __init__ and remove interpolation check from gpu_simulate

kartikmandar committed:
remove pragma: no cover lines from code

kartikmandar committed:
removed redundant checks from simulate and add use=gpu to the coord_mgr

kartikmandar committed:
remove duplication of flux_slice call

kartikmandar committed:
raising error if reshaping fail than a warning

kartikmandar committed:
refactor: rename simulate modules, simplify GPU beams, update imports

- Rename cpu_simulate.py → cpu.py and gpu_simulate.py → gpu.py for
  consistency with matvis naming convention
- Simplify GPUBeamEvaluator to inherit directly from matvis
  GPUBeamInterpolator, removing ~400 lines of duplicate logic
- Update all imports and test files for new module names
- Refactor CPUBeamEvaluator for cleaner matvis integration

kartikmandar committed:
refactor: remove Type 3 NUFFT fallback code, require cufinufft>=2.2.0

Remove _test_native_type3_support, HAS_NATIVE_TYPE3_SUPPORT flag, and
all plan-based fallback functions. cufinufft stable releases (2.2.0+)
have native Type 3 support, making the fallback unnecessary. This
removes ~200 lines of duplicate code paths and simplifies error handling.

kartikmandar committed:
fix: remove sleep from cleanup_gpu_memory, clean up gpu/utils.py

Remove arbitrary 0.1s sleep from cleanup_gpu_memory() — deviceSynchronize
is sufficient. Replace bare except with debug-level logging. Remove
"use CPU" error messages from execute_with_retry(). Remove unused imports
(time, Optional). isinstance check for cupy arrays was already removed.

kartikmandar committed:
refactor: remove GPU capacity limits, fix coord_mgr initialization order

Remove artificial source-count hard limits that told users to "use CPU
backend". GPU should handle any dataset size through chunking, following
matvis's pattern.

Compute optimal chunk_size via optimize_chunking() BEFORE creating the
CoordinateRotation manager, and pass chunk_size directly to the
constructor instead of mutating it after creation.

kartikmandar committed:
refactor: clean up gpu.py — BeamInterface type, remove dtype cast and memory cleanup

- Always expect BeamInterface in _evaluate_vis_chunk, update type hint
- Handle both BeamInterface and bare UVBeam in simulate() for
  backwards compatibility, wrapping UVBeam in BeamInterface
- Remove unnecessary dtype cast on beam output (beam evaluator
  should return correct dtype via precision parameter)
- Remove periodic mempool.free_all_blocks() with magic numbers
  (200000 sources, every 10 batches) — proper chunking handles memory

kartikmandar committed:
refactor: move beam_list init into __init__, use dict merge in wrapper

- Move beam_list/beam_idx initialization from create_beam_evaluator()
  factory into GPUBeamEvaluator.__init__() (CPUBeamEvaluator already
  had it). Factory now just returns the constructor result.
- Use dict merge pattern (common_params | backend_params) for cleaner
  separation of common vs backend-specific parameters in simulate_vis().

kartikmandar committed:
docs: document memory_manager relationship with matvis get_desired_chunks

kartikmandar committed:
refactor: remove unused GPU capability flags from beams.py

Remove HAS_GPU_BESSEL, HAS_GPU_MAP_COORDS, and their associated imports
(gpu_j1, gpu_map_coordinates). These are no longer used since
GPUBeamEvaluator delegates to matvis's GPUBeamInterpolator which handles
its own GPU interpolation internally.

kartikmandar committed:
docs: add TODO for moving GPU data transfers into components

Document the intention to follow matvis's pattern where each component
manages its own GPU state. The actual refactor is deferred as it
requires introducing a GPU data holder and changing function signatures.

kartikmandar committed:
refactor: extract shared CPU/GPU setup into prepare_simulation_inputs()

Add prepare_simulation_inputs() helper to core/simulate.py that handles
the common setup logic duplicated between CPU and GPU backends: dtype
resolution, eps defaults, baseline computation, antenna rotation,
coplanarity check, and Time conversion.

Both CPUSimulationEngine and GPUSimulationEngine now call this helper
instead of duplicating ~60 lines of identical setup code. CPU retains
its griddability check as a backend-specific path.

kartikmandar committed:
fix: update test to match new CUDA error message format

kartikmandar committed:
refactor: address remaining PR review feedback from Steven

- Update find_integer_multiplier docstring to suggested wording
- Remove redundant beam UVBeam/BeamInterface handling in GPU simulate()
- Replace TODO with rationale for GPU data transfers location
- Clarify GPUMemoryManager vs matvis get_desired_chunks relationship
- Document cleanup_gpu_memory purpose (CuPy pool flush for OOM retry)

steven-murray commented:
@kartikmandar it looks like we need many more tests to get the patch coverage to (close to) 100%. This is super important, especially on a big PR like this where a lot can go wrong.
